ouroboros-consensus-cardano/tools/db-analyser/Main.hs

```diff
@@ -0,0 +1,330 @@
+{-# LANGUAGE FlexibleContexts    #-}
+{-# LANGUAGE OverloadedStrings   #-}
+{-# LANGUAGE RankNTypes          #-}
+{-# LANGUAGE RecordWildCards     #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications    #-}
+
+module Main (main) where
+
+import           Control.Monad.Except
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Lazy as BL
+import           Data.Foldable (asum)
+import           Data.IORef
+import           Data.List (foldl', intercalate)
+import           Data.Proxy (Proxy (..))
+import qualified Data.Text as Text
+import           Data.Word
+import           GHC.Int (Int64)
+import           GHC.Natural (Natural)
+import           Options.Applicative
+
+import           Cardano.Slotting.Slot
+
+import           Cardano.Crypto.ProtocolMagic
+
+import           Ouroboros.Network.Block (HasHeader (..), HeaderHash,
+                     genesisPoint)
+
+import           Ouroboros.Consensus.Block
+import           Ouroboros.Consensus.Config
+import           Ouroboros.Consensus.Node.DbMarker
+import           Ouroboros.Consensus.Node.Run
+import           Ouroboros.Consensus.Util.ResourceRegistry
+
+import           Ouroboros.Consensus.Storage.ChainDB.API (BlockComponent (..),
+                     StreamFrom (..), StreamTo (..))
+import           Ouroboros.Consensus.Storage.ChainDB.Impl.ImmDB hiding
+                     (withImmDB)
+import qualified Ouroboros.Consensus.Storage.ChainDB.Impl.ImmDB as ImmDB
+                     (withImmDB)
+import qualified Ouroboros.Consensus.Storage.ImmutableDB.API as ImmDB
+
+import           Ouroboros.Consensus.Byron.Ledger (ByronBlock)
+import           Ouroboros.Consensus.Cardano.Block (CardanoBlock)
+import           Ouroboros.Consensus.Shelley.Ledger.Block (ShelleyBlock (..))
+import           Ouroboros.Consensus.Shelley.Protocol.Crypto
+                     (TPraosStandardCrypto)
+
+import           Analysis hiding (blockTxs, countTxOutputs)
+import qualified Analysis
+
+main :: IO ()
+main = do
+    cmdLine@CmdLine{..} <- getCmdLine
+    case clBlockType of
+      Just Byron   -> analyse cmdLine (Proxy @ByronBlock)
+      Just Shelley -> analyse cmdLine (Proxy @(ShelleyBlock TPraosStandardCrypto))
+      Just Cardano -> analyse cmdLine (Proxy @(CardanoBlock TPraosStandardCrypto))
+      Nothing -> do
+        -- check the dbmarker of the db if the block type is not specified.
+        protocolMagicId <- readDBMarker clImmDB
+        case protocolMagicId of
+          764824073  -> analyse cmdLine (Proxy @ByronBlock)
+          1097911063 -> analyse cmdLine (Proxy @ByronBlock)
+          42         -> analyse cmdLine (Proxy @(ShelleyBlock TPraosStandardCrypto))
+          _          -> error $ "unsupported protocolMagicId: " ++ show protocolMagicId
+
+readDBMarker :: FilePath -> IO Word32
+readDBMarker dbPath = do
+    bs <- BS.readFile markerPath
+    protocolMagicId <- runExceptT $ dbMarkerParse markerPath bs
```

Comment:
That's better :+1: 

I would still let it return a `ProtocolMagicId`, as it makes it clear what we're returning, which is not the case for `Word32`. You can then unwrap it in `main`.